### PR TITLE
Unify error message for better Catalyst support

### PR DESCRIPTION
**Pull request by multiphaseCFD**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**

Unify error messages of shot measurement related unsupported observables to better Catalyst.

**Description of the Change:**

For the moment, shot measurements support `Paulis`, `Identity` and `Hadamard` obs and `TensorProd/Hamiltonian` obs created based on above mentioned named `obs`. Error messages are thrown inside each shot measurement method. Now error messages are obs based and it will support Catalyst better.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by AmintorDusko** [comment]
@multiphaseCFD, could you please fill-out the PR description?

File: pennylane_lightning/core/src/measurements/MeasurementsBase.hpp [hunk 1]
```diff
@@ -133,18 +133,9 @@ template <class StateVectorT, class Derived> class MeasurementsBase {
...
-            PL_ABORT("For SparseHamiltonian Observables, expval calculation is "
-                     "not supported by shots");
-        } else if (obs.getObsName().find("Hermitian") != std::string::npos) {
-            // TODO support. This support requires an additional method to solve
-            // eigenpair and unitary matrices, and the results of eigenpair and
-            // unitary matrices data need to be added to the Hermitian class and
-            // public methods are need to access eigen values. Note the
-            // assumption that eigen values are -1 and 1 in the
-            // `measurement_with_sample` method should be updated as well.
-            PL_ABORT("For Hermitian Observables, expval calculation is not "
```
Comment (vincentmr) [inline pennylane_lightning/core/src/measurements/MeasurementsBase.hpp:146]: Is this just caught in Observables now?

**Comment by multiphaseCFD** [comment]
> @multiphaseCFD, could you please fill-out the PR description?

Thanks @AmintorDusko . Done!

File: pennylane_lightning/core/src/measurements/MeasurementsBase.hpp (hunk 1 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/measurements/MeasurementsBase.hpp:146]: Yea, it is caught by the `applyInPlaceShots()` in Observables.

**Review by vincentmr — APPROVED** [review]
Looks good to me, cheers @multiphaseCFD !

**Review by AmintorDusko — APPROVED** [review]
Nice work @multiphaseCFD! Thank you for that!

**Comment by maliasadi** [comment]
Hey Shuli, Would you like to include the patch to address the following issue as well?

`var(TensorProd, num_shots=5000)` using the C++ API in both Lightning and LightningKokkos returns zero for the following circuit:

``` python
@qml.qnode(dev)
def circuit():
    qml.PauliX(0)
    qml.PauliY(1)
    qml.Hadamard(2)
    qml.PauliZ(3)
    return qml.var(qml.PauliX(2) @ qml.PauliY(1) @ qml.PauliZ(0))

>>> circuit()
0.99983616
```

**Comment by multiphaseCFD** [comment]
> ` [565387d](/PennyLaneAI/pennylane-lightning/pull/577/commits/565387d46aa08c47e5b2d843a9e888f3e56ba266)`

Hey @maliasadi , just tested the circuit you mentioned above with [565387d](https://github.com/PennyLaneAI/pennylane-lightning/pull/577/commits/565387d46aa08c47e5b2d843a9e888f3e56ba266) and tests passed in C++ backend.

File: pennylane_lightning/core/src/observables/Observables.hpp [hunk 2]
```diff
@@ -20,6 +20,8 @@
 #include <unordered_set>
 #include <vector>
 
+#include <iostream>
+
```
Comment (maliasadi) [inline pennylane_lightning/core/src/observables/Observables.hpp:24]:
```suggestion
```

File: pennylane_lightning/core/src/observables/Observables.hpp [hunk 3]
```diff
@@ -189,15 +191,13 @@ class NamedObsBase : public Observable<StateVectorT> {
...
         } else {
-            PL_ABORT("Provided NamedObs does not supported for shots "
-                     "calculation. Supported NamedObs are PauliX, PauliY, "
-                     "PauliZ, Identity and Hadamard.");
+            PL_ABORT("Provided NamedObs does not support shot measurement.");
         }
 
         if (obs_name_ == "Identity") {
-            eigenValues.push_back({1, 1});
+            eigenValues.push_back(std::vector<PrecisionT>{1, 1});
```
Comment (maliasadi) [inline pennylane_lightning/core/src/observables/Observables.hpp:198]:
```suggestion
            eigenValues.push_back({1, 1});
```

File: pennylane_lightning/core/src/observables/Observables.hpp [hunk 4]
```diff
@@ -189,15 +191,13 @@ class NamedObsBase : public Observable<StateVectorT> {
...
         } else {
-            eigenValues.push_back({1, -1});
+            eigenValues.push_back(std::vector<PrecisionT>{1, -1});
```
Comment (maliasadi) [inline pennylane_lightning/core/src/observables/Observables.hpp:200]:
```suggestion
            eigenValues.push_back({1, -1});
```

**Review by maliasadi — APPROVED** [review]

File: pennylane_lightning/core/src/observables/Observables.hpp [hunk 5]
```diff
@@ -300,8 +301,15 @@ class TensorProdObsBase : public Observable<StateVectorT> {
      */
     template <typename... Ts>
     explicit TensorProdObsBase(Ts &&...arg) : obs_{std::forward<Ts>(arg)...} {
-        std::unordered_set<size_t> wires;
+        if (obs_.size() == 1 &&
```
Comment (maliasadi) [inline pennylane_lightning/core/src/observables/Observables.hpp:304]: @multiphaseCFD

File: pennylane_lightning/core/src/observables/tests/Test_Observables.cpp [hunk 6]
```diff
@@ -277,6 +278,18 @@ template <typename TypeList> void testTensorProdObsBase() {
             REQUIRE_NOTHROW(TensorProdObsT::create({ob1, ob2}));
         }
 
+        DYNAMIC_SECTION("Constructing an invalid TensorProd(TensorProd) - "
```
Comment (maliasadi) [inline pennylane_lightning/core/src/observables/tests/Test_Observables.cpp:281]: @multiphaseCFD

File: pennylane_lightning/core/src/observables/Observables.hpp (hunk 5 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/observables/Observables.hpp:304]: Nice one!

File: pennylane_lightning/core/src/observables/tests/Test_Observables.cpp (hunk 6 above)
Comment (multiphaseCFD) [reply pennylane_lightning/core/src/observables/tests/Test_Observables.cpp:281]: Nice add!

**Pull request merged**